### PR TITLE
16 parse the configuration file

### PR DESCRIPTION
**Pull request by AxelF44**
I've implemented the parsing of the configuration file,

I've added All light plugins !

### CODE REVIEW

File: include/Camera/Camera.hpp [hunk 1]
```diff
@@ -64,9 +64,47 @@ class Raytracer::Camera {
...
+         * @brief Get Fov of Camera
+         *
+         */
+        double getFov(void);
+
+        /**
+         * @brief Get Rotation of Camera
+         *
+         */
+        Math::Vector3D getRotation(void);
```
Comment (Njord201) [inline include/Camera/Camera.hpp:78]: Const

File: include/Camera/Camera.hpp [hunk 2]
```diff
@@ -64,9 +64,47 @@ class Raytracer::Camera {
...
+
+        /**
+         * @brief Set origin of Camera
+         *
+         */
+        void setOrigin(Math::Point3D origin);
```
Comment (Njord201) [inline include/Camera/Camera.hpp:84]: Describe the parameters in doxygen documentation.

File: include/Camera/Camera.hpp [hunk 3]
```diff
@@ -64,9 +64,47 @@ class Raytracer::Camera {
...
+
+        /**
+         * @brief Set Screen of Camera
+         *
+         */
+        void setScreen(Raytracer::Rectangle3D screen);
```
Comment (Njord201) [inline include/Camera/Camera.hpp:90]: Describes parameters.

File: include/Camera/Camera.hpp [hunk 4]
```diff
@@ -64,9 +64,47 @@ class Raytracer::Camera {
...
+
+        /**
+         * @brief Set Fov of Camera
+         *
+         */
+        void setFov(double fov);
```
Comment (Njord201) [inline include/Camera/Camera.hpp:96]: Describe parameters

File: include/Camera/Camera.hpp [hunk 5]
```diff
@@ -64,9 +64,47 @@ class Raytracer::Camera {
...
+
+        /**
+         * @brief Set Rotation of Camera
+         *
+         */
+        void setRotation(Math::Vector3D rotation);
```
Comment (Njord201) [inline include/Camera/Camera.hpp:102]: Describe parameters

File: include/Lights/Ambient.hpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+         * @brief Construct a new Ambient object
+         *
+         */
+        Ambient();
+
+        /**
+         * @brief Construct a new Ambient object
+         *
+         */
+        Ambient(double multiplier, double diffuseMultiplier);
```
Comment (Njord201) [inline include/Lights/Ambient.hpp:32]: Describe parameters

File: include/Lights/Ambient.hpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+         * @brief Get the Multiplier number of ambient light
+         * 
+         * @return double 
+         */
+        double getMultiplier(void);
+
+        /**
+         * @brief Set the Multiplier number of ambient light
+         */
+        void setMultiplier(double multiplier);
```
Comment (Njord201) [inline include/Lights/Ambient.hpp:50]: Describe parameters

File: include/Lights/Ambient.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        /**
+         * @brief Get the Diffuse Multiplier number of ambient light
+         * 
+         * @return double 
+         */
+        double getDiffuseMultiplier(void);
```
Comment (Njord201) [inline include/Lights/Ambient.hpp:57]: Const

File: include/Lights/Ambient.hpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+        /**
+         * @brief Set the Diffuse Multiplier number of ambient light
+         */
+        void setDiffuseMultiplier(double diffuseMultiplier);
```
Comment (Njord201) [inline include/Lights/Ambient.hpp:62]: Describe parameters

File: include/Lights/Ambient.hpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+        /**
+         * @brief Get type of Light
+         * @return The type of the light
+         */
+        Light::LightType getType(void) override;
```
Comment (Njord201) [inline include/Lights/Ambient.hpp:68]: Const

File: include/Lights/Directional.hpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+         * @brief Construct a new Directional object
+         *
+         */
+        Directional();
+
+        /**
+         * @brief Construct a new Directional object
+         *
+         */
+        Directional(Math::Point3D position, Math::Vector3D direction, double diffuseMultiplier);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:33]: Describe parameters

File: include/Lights/Directional.hpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+         *
+         */
+        ~Directional();
+
+        /**
+         * @brief Get the Position number of Point light
+         * 
+         * @return double 
+         */
+        Math::Point3D getPosition(void);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:46]: Const

File: include/Lights/Directional.hpp [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+        /**
+         * @brief Set the Position number of Point light
+         */
+        void setPosition(Math::Point3D position);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:51]: Parameters

File: include/Lights/Directional.hpp [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+        /**
+         * @brief Get the Direction number of Point light
+         * 
+         * @return double 
+         */
+        Math::Vector3D getDirection(void);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:58]: Const

File: include/Lights/Directional.hpp [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+        /**
+         * @brief Set the Direction number of Point light
+         */
+        void setDirection(Math::Vector3D direction);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:63]: Describe parameters

File: include/Lights/Directional.hpp [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+        /**
+         * @brief Get the Diffuse Multiplier number of Point light
+         * 
+         * @return double 
+         */
+        double getDiffuseMultiplier(void);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:70]: Const

File: include/Lights/Directional.hpp [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+        /**
+         * @brief Set the Diffuse Multiplier number of Point light
+         */
+        void setDiffuseMultiplier(double diffuseMultiplier);
```
Comment (Njord201) [inline include/Lights/Directional.hpp:75]: Describe parameters

File: include/Lights/Directional.hpp [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+
+        /**
+         * @brief Get type of Light
+         * @return The type of the light
+         */
+        Light::LightType getType(void) override;
```
Comment (Njord201) [inline include/Lights/Directional.hpp:81]: Const

File: include/Lights/ILight.hpp [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+             * @brief Destroy the ILight object
+             *
+             */
+            virtual ~ILight() = default;
+
+            /**
+             * @brief Get type of Light
+             * @return The type of the light
+             */
+            virtual Light::LightType getType(void) = 0;
```
Comment (Njord201) [inline include/Lights/ILight.hpp:35]: Const

File: include/Lights/LightsContainer.hpp [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+             *
+             */
+            void clear();
+
+            /**
+             * @brief Get the Lights List object
+             * 
+             * @return std::vector<std::shared_ptr<Light::ILight>> 
+             */
+            std::vector<std::shared_ptr<Light::ILight>> getLightsList(void);
```
Comment (Njord201) [inline include/Lights/LightsContainer.hpp:52]: Const

File: include/Lights/Point.hpp [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+         * @brief Construct a new Point object
+         *
+         */
+        Point();
+
+        /**
+         * @brief Construct a new Point object
+         *
+         */
+        Point(Math::Point3D position, double diffuseMultiplier);
```
Comment (Njord201) [inline include/Lights/Point.hpp:33]: Describe parameters

File: include/Lights/Point.hpp [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+         *
+         */
+        ~Point();
+
+        /**
+         * @brief Get the Position number of Point light
+         * 
+         * @return double 
+         */
+        Math::Point3D getPosition(void);
```
Comment (Njord201) [inline include/Lights/Point.hpp:46]: Const

File: include/Lights/Point.hpp [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+         * @brief Set the Position number of Point light
+         */
+        void setPosition(Math::Point3D position);
+
+        /**
+         * @brief Get the Diffuse Multiplier number of Point light
+         * 
+         * @return double 
+         */
+        double getDiffuseMultiplier(void);
```
Comment (Njord201) [inline include/Lights/Point.hpp:58]: Const

File: include/Lights/Point.hpp [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+         * @brief Get the Position number of Point light
+         * 
+         * @return double 
+         */
+        Math::Point3D getPosition(void);
+
+        /**
+         * @brief Set the Position number of Point light
+         */
+        void setPosition(Math::Point3D position);
```
Comment (Njord201) [inline include/Lights/Point.hpp:51]: Describe parameters

File: include/Lights/Point.hpp [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+         * @brief Get the Diffuse Multiplier number of Point light
+         * 
+         * @return double 
+         */
+        double getDiffuseMultiplier(void);
+
+        /**
+         * @brief Set the Diffuse Multiplier number of Point light
+         */
+        void setDiffuseMultiplier(double diffuseMultiplier);
```
Comment (Njord201) [inline include/Lights/Point.hpp:63]: Describe parameters

File: include/Lights/Point.hpp [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+
+        /**
+         * @brief Get type of Light
+         * @return The type of the light
+         */
+        Light::LightType getType(void) override;
```
Comment (Njord201) [inline include/Lights/Point.hpp:69]: Const

File: include/Materials/FlatColor.hpp [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+         * @brief Destruct a FlatColor object
+         *
+         */
+        ~FlatColor();
+
+        /**
+         * @brief Get type of Material
+         * @return The type of the material
+         */
+        Material::MaterialType getType(void) override;
```
Comment (Njord201) [inline include/Materials/FlatColor.hpp:40]: Const

File: include/Materials/FlatColor.hpp [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+         * @brief Set the new B color
+         * @param b new g color
+         */
+        void setB(int b);
+
+        /**
+         * @brief Get the R color
+         * @return R color
+         */
+        int getR(void);
```
Comment (Njord201) [inline include/Materials/FlatColor.hpp:64]: Const

File: include/Materials/FlatColor.hpp [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+
+        /**
+         * @brief Get the G color
+         * @return G color
+         */
+        int getG(void);
```
Comment (Njord201) [inline include/Materials/FlatColor.hpp:70]: Const

File: include/Materials/FlatColor.hpp [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+        /**
+         * @brief Get the B color
+         * @return B color
+         */
+        int getB(void);
```
Comment (Njord201) [inline include/Materials/FlatColor.hpp:76]: Const

File: include/Materials/IMaterial.hpp [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+         * @brief Destroy the IMaterial object
+         *
+         */
+        virtual ~IMaterial() = default;
+
+        /**
+         * @brief Get type of Material
+         * @return The type of the material
+         */
+        virtual MaterialType getType(void) = 0;
```
Comment (Njord201) [inline include/Materials/IMaterial.hpp:32]: Const

File: include/Parser/DLLoader.hpp [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+         */
+        ~DLLoader();
+
+        /**
+         * @brief Get the Instance object
+         * @param functionName
+         * @return T
+         */
+        template <typename T>
+        T getInstance(const std::string functionName) const
```
Comment (Njord201) [inline include/Parser/DLLoader.hpp:44]: Don't do logic in hpp files, move getInstance in a cpp

File: include/Parser/Factory.hpp [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+
+        /// @brief Register a type component
+        /// @param type Component to register
+        /// @param creator Function to create the component
+        void registerLightsComponent(const std::string& type, LightsCreator creator);
+    
+    private:
+        std::unordered_map<std::string, PrimitivesCreator>   _componentPrimitivesList;
+        std::unordered_map<std::string, LightsCreator>       _componentLightsList;
+        std::vector<std::shared_ptr<DLLoader>> _libraryLoader;
```
Comment (Njord201) [inline include/Parser/Factory.hpp:71]: Try to keep logic of \t

File: include/Parser/Scene.hpp [hunk 34]
```diff
@@ -0,0 +1,110 @@
...
+
+    public:
+
+        using PrimitivesCreator = std::function<std::unique_ptr<Primitive::IPrimitive>()>;
+
+        /**
+         * @brief Construct a new Scene object
+         *
+         */
+        Scene(std::string filePath);
```
Comment (Njord201) [inline include/Parser/Scene.hpp:39]: Describe parameters

File: include/Parser/Scene.hpp [hunk 35]
```diff
@@ -0,0 +1,110 @@
...
+
+        /**
+         * @brief Destruct a Scene object
+         *
+         */
+        ~Scene();
+
+        class ParserException : public std::exception {
```
Comment (Njord201) [inline include/Parser/Scene.hpp:47]: Try to move the declaration after the Scene class

File: include/Parser/Scene.hpp [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+            private:
+                std::string message;
+        };
+
+        /**
+         * @brief Get the Camera object
+         * 
+         * @return Camera Object
+         */
+        Raytracer::Camera &getCamera(void);
```
Comment (Njord201) [inline include/Parser/Scene.hpp:68]: Const

File: include/Parser/Scene.hpp [hunk 37]
```diff
@@ -0,0 +1,110 @@
...
+        
+        /**
+         * @brief Get the Primitives object
+         * 
+         * @return Primitive::PrimitivesContainer 
+         */
+        Primitive::PrimitivesContainer getPrimitives(void);
```
Comment (Njord201) [inline include/Parser/Scene.hpp:75]: Const

File: include/Parser/Scene.hpp [hunk 38]
```diff
@@ -0,0 +1,110 @@
...
+
+        /**
+         * @brief Get the Lights object
+         * 
+         * @return Light::LightsContainer
+         */
+        Light::LightsContainer getLights(void);
```
Comment (Njord201) [inline include/Parser/Scene.hpp:82]: Const

File: include/Primitives/Sphere.hpp [hunk 39]
```diff
@@ -25,13 +30,13 @@ namespace Primitive {
              * @param origin center of the sphere
              * @param radius of the sphere
              */
-            Sphere(const Math::Point3D& origin, double radius);
+            Sphere(const Math::Point3D& origin, double radius, std::shared_ptr<Material::IMaterial> material);
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:33]: Desribe the param added

File: include/Primitives/Sphere.hpp [hunk 40]
```diff
@@ -41,9 +46,45 @@ namespace Primitive {
...
+             * @brief Set the Material
+             * @param material New material to set
+             */
+            void setMaterial(std::shared_ptr<Material::IMaterial> material);
+
+            /**
+             * @brief Get the Origin object
+             * @return Origin of sphere
+             */
+            Math::Point3D getOrigin();
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:71]: Const

File: include/Primitives/Sphere.hpp [hunk 41]
```diff
@@ -41,9 +46,45 @@ namespace Primitive {
...
+            
+            /**
+             * @brief Get the Origin object
+             * @return Radius of sphere
+             */
+            double getRadius();
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:77]: Const

File: include/Primitives/Sphere.hpp [hunk 42]
```diff
@@ -41,9 +46,45 @@ namespace Primitive {
...
 
-            Math::Point3D _origin;
-            double _radius;
+            /**
+             * @brief Get the Material object
+             * @return Material of sphere
+             */
+            std::shared_ptr<Material::IMaterial> getMaterial();
```
Comment (Njord201) [inline include/Primitives/Sphere.hpp:83]: Const

File: src/Core/Camera/Camera.cpp [hunk 43]
```diff
@@ -43,3 +43,33 @@ Raytracer::Rectangle3D Raytracer::Camera::getScreen(void)
 {
     return this->_screen;
 }
+
+double Raytracer::Camera::getFov(void)
```
Comment (Njord201) [inline src/Core/Camera/Camera.cpp:47]: Const

File: src/Core/Camera/Camera.cpp [hunk 44]
```diff
@@ -43,3 +43,33 @@ Raytracer::Rectangle3D Raytracer::Camera::getScreen(void)
...
+{
+    return this->_fov;
+}
+
+Math::Vector3D Raytracer::Camera::getRotation(void)
```
Comment (Njord201) [inline src/Core/Camera/Camera.cpp:52]: Const

File: src/Core/Lights/LightsContainer.cpp [hunk 45]
```diff
@@ -0,0 +1,27 @@
...
+*/
+
+#include "Lights/LightsContainer.hpp"
+
+void Light::LightsContainer::add(std::shared_ptr<Light::ILight> primitive)
+{
+    _Lights.push_back(primitive);
+}
+
+Light::LightsContainer::~LightsContainer()
```
Comment (Njord201) [inline src/Core/Lights/LightsContainer.cpp:15]: Put the destructor in default

File: src/Core/Lights/LightsContainer.cpp [hunk 46]
```diff
@@ -0,0 +1,27 @@
...
+{
+}
+
+void Light::LightsContainer::clear()
+{
+    _Lights.clear();
+}
+
+std::vector<std::shared_ptr<Light::ILight>> Light::LightsContainer::getLightsList(void)
```
Comment (Njord201) [inline src/Core/Lights/LightsContainer.cpp:24]: Const

File: src/Core/Materials/FlatColor.cpp [hunk 47]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+FlatColor::FlatColor(int r, int g, int b)
+{
+    this->_r = r;
+    this->_g = g;
+    this->_b = b;
+}
+
+FlatColor::~FlatColor()
```
Comment (Njord201) [inline src/Core/Materials/FlatColor.cpp:25]: Put the destuctor in default

File: src/Core/Parser/Factory.cpp [hunk 48]
```diff
@@ -0,0 +1,77 @@
...
+        return sharedPtrPoint;
+    });
+    this->registerLightsComponent("directional", [directionalLoader]() -> std::shared_ptr<Light::ILight> {
+        Light::ILight *directional = directionalLoader->getInstance<Light::ILight *>("getDirectionalInstance");
+        std::shared_ptr<Light::ILight> sharedPtrDirectional(directional);
+        return sharedPtrDirectional;
+    });
+}
+
+Raytracer::Factory::~Factory()
```
Comment (Njord201) [inline src/Core/Parser/Factory.cpp:43]: Put the destructor in default

File: src/Core/Primitives/PrimitivesContainer.cpp [hunk 49]
```diff
@@ -24,4 +27,11 @@ Math::Point3D Primitive::PrimitivesContainer::hitPoint(const Raytracer::Ray& ray
         _primitives[i]->hitPoint(ray);
         // TODO : implement the hitpoint logic
     }
+    Math::Point3D temp(0, 0, 0);
+    return temp;
+}
+
+std::vector<std::shared_ptr<Primitive::IPrimitive>> Primitive::PrimitivesContainer::getPrimitivesList(void)
```
Comment (Njord201) [inline src/Core/Primitives/PrimitivesContainer.cpp:34]: Const

File: src/Plugins/Lights/Point/Point.cpp [hunk 50]
```diff
@@ -0,0 +1,50 @@
...
+{
+    return this->_diffuseMultiplier;
+}
+
+void Light::Point::setDiffuseMultiplier(double diffuseMultiplier)
+{
+    this->_diffuseMultiplier = diffuseMultiplier;
+}
+
+Light::LightType Light::Point::getType(void)
```
Comment (Njord201) [inline src/Plugins/Lights/Point/Point.cpp:47]: Const

File: src/Plugins/Primitives/Sphere/Sphere.cpp [hunk 51]
```diff
@@ -36,3 +41,33 @@ Math::Point3D Primitive::Sphere::hitPoint(const Raytracer::Ray& r) const
...
+{
+    this->_material = material;
+}
+
+Math::Point3D Primitive::Sphere::getOrigin()
+{
+    return this->_origin;
+}
+
+double Primitive::Sphere::getRadius()
```
Comment (Njord201) [inline src/Plugins/Primitives/Sphere/Sphere.cpp:65]: Const

**Review by Njord201 — CHANGES_REQUESTED** [review]
Some const, documentation and deafult forgotten.  But on the whole, it's good.

File: include/Parser/DLLoader.hpp (hunk 32 above)
Comment (mathieurobert1) [reply include/Parser/DLLoader.hpp:44]: No Axel is right, a template function is describe in an hpp file.

File: include/Primitives/PrimitivesContainer.hpp [hunk 52]
```diff
@@ -7,10 +7,11 @@
 
 #pragma once
 
-#include "IPrimitive.hpp"
-
 #include <vector>
 #include <memory>
+#include <iostream>
```
Comment (mathieurobert1) [inline include/Primitives/PrimitivesContainer.hpp:12]: useless include

**Review by Njord201 — CHANGES_REQUESTED** [review]
There is still some unresolved conversations in the file.cpp file. Add const and descriptions and you're done.

**Review by mathieurobert1 — APPROVED** [review]

**Review by Njord201 — APPROVED** [review]
:+1:

**Pull request merged**